### ISSUE TITLE
Inference issue between the diamond syntax and pattern matching (switch on objects)

### ISSUE THREAD

**Issue by forax** [eclipse-jdt/eclipse.jdt.core#1788]
Hello,
the following code does not compile while it compiles with javac 21.

"Cannot infer type arguments for Entry<>" at line 24.

--- 
```
import java.util.Map;

public class EclipseSwitchInferenceBug {
  
  final static class Entry<K,V> {
    K key;
    V value;
    
    public Entry(K key, V value) {
      this.key = key;
      this.value = value;
    }
    
    public K key() { return key; }
    public V value() { return value; }
    public void value(V value) { this.value = value; }
  }
  
  sealed interface I<V> {}
  record A<V>(V value) implements I<V> {}
  record B<V>(V value) implements I<V> {}
  
  private static <K, V> Entry<K,V> foo(Entry<K, I<V>> entry) {
    return new Entry<>(entry.key(), switch (entry.value()) {
      case A<V>(V value) -> {
        entry.value(new B<>(value));
        yield value;
      }
      case B<V>(V value) -> value;
    });
  }
}
```

**Comment by snjeza** [comment]
It works fine in JDT master. You can try https://download.eclipse.org/eclipse/downloads/drops4/I20231223-1800/

**Comment by srikanth-sankaran** [comment]
Seems to have been fixed by the commit d4e6217d27112cd5c270638de416ec04bc8a8d46 made for https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1726

**Comment by srikanth-sankaran** [comment]
> Seems to have been fixed by the commit [d4e6217](https://github.com/eclipse-jdt/eclipse.jdt.core/commit/d4e6217d27112cd5c270638de416ec04bc8a8d46) made for #1726

Yep, I remember that fix and can see how that can address this problem. The parse tree for the return expression was messed up before the fix
